CrossPare/src/main/java/de/ugoe/cs/cpdp/dataprocessing/SMOTUNED.java

```diff
@@ -0,0 +1,638 @@
+            this.smotePercentage = smotePercentage;
+            this.minkowskiOrder = minkowskiOrder;
+            this.undersampleRatio = undersampleRatio;
+        }
+
+        public SMOTE(SMOTE other) {
+            this.numNeighbors = other.numNeighbors;
+            this.smotePercentage = other.smotePercentage;
+            this.minkowskiOrder = other.minkowskiOrder;
+            this.undersampleRatio = other.undersampleRatio;
+        }
+
+        /**
+         * @return the number of neighbors
+         */
+        public int getK() { return numNeighbors; }
+
+        /**
+         * @return the number of synthetic examples / desired ratio between minority and majority instances
+         */
+        public int getM() { return smotePercentage; }
+
+        /**
+         * @return the Minkowski order
+         */
+        public double getMinkowskiOrder() { return minkowskiOrder; }
+
+        /**
+         * Applies the SMOTE algorithm to the data set.
+         * <p>
+         * Additionally introduces two index attributes to the instances.
+         * In the case of an unchanged minority or majority element the 'instance_index' is the index
+         * in the original data set and the 'synthetic_index' is set to -1.
+         * For synthesized elements the 'instance_index' and 'synthetic_index' are the indices of its two
+         * root elements in the original data set.
+         *
+         * @param data original training data
+         * @return training data manipulated by SMOTE with additional indices
+         */
+        private Instances applySMOTE(Instances data){
+            Random rand = new Random();
+
+            Instances indexed = new Instances(data);
+            indexed.insertAttributeAt(new Attribute("instance_index"), 0);
+            for (int i=0; i<indexed.size(); i++) {
+                indexed.get(i).setValue(0, i);
+            }
+            indexed.insertAttributeAt(new Attribute("synthetic_index"), 1);
+            for (int i=0; i<indexed.size(); i++) {
+                indexed.get(i).setValue(1, -1);
+            }
+            Instances minority = new Instances(indexed, 0);
+            Instances majority = new Instances(indexed, 0);
+            for(int i=0; i<indexed.size(); i++){
+                if (indexed.get(i).classValue() > 0.0) {
+                    minority.add(indexed.get(i));
+                } else {
+                    majority.add(indexed.get(i));
+                }
+            }
+            final int sizeMinority = minority.size();
+            if (sizeMinority < 2){
+                throw new RuntimeException("error applying SMOTE: SMOTE needs at least 2 defective Instances.");
+            }
+            // reduce amount of neighbors if numNeighbors bigger than minority size
+            if (this.numNeighbors >= sizeMinority){
+                this.numNeighbors = sizeMinority - 1;
+            }
+            ArrayList<Integer> synthList = new ArrayList<Integer>();
```

Comment:
general style hint: 
avoid `ArrayList` and `LinkedList` as types, always use `List` instead. this makes the code easier to maintain and switch types (same for sets, etc).